include/circt/Dialect/SV/SVPasses.td

```diff
@@ -0,0 +1,47 @@
+//===-- SVPasses.td - SV pass definition file --------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains definitions for passes that work on the SV dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_SV_SVPASSES
+#define CIRCT_DIALECT_SV_SVPASSES
+
+include "mlir/Pass/PassBase.td"
+
+def AlwaysFusion : Pass<"sv-always-fusion"> {
```

Comment:
Out of curiosity, why not make this be specific to `rtl.module`?  While it is true that this adds a dependency on the RTL dialect, I don't think that really matters (the bulk of RTL will be split out to comb at some point anyway).  This would also make this thing run faster, because it will only be invoked on the top level module, instead of being invoked on every operation individually on the module.

Such an implementation would have to implement a RPO walk of the AST to visit things in the right order, but I think that is ok and will be useful for other reasons.  Do you have any objection with me or you refactoring this in this structure?

I would also like to generalize this into a "tidy-rtl" pass, which can do other simplifications as well - merge ifdefs and ifs, form casex's, etc.

WDYT?

Reply:
> why not make this be specific to rtl.module?

I wrote the pass so that it could be invoked on any top level module-like operation. It seemed like it didn't really cost me anything to make it op-agnostic, but maybe I'm wrong.  The pass is invoked like this:
```c++
pm.addNestedPass<rtl::RTLModuleOp>(sv::createAlwaysFusionPass());
```
The pass will scan all graph regions directly under the module for `sv.alwaysff` ops and then merge them.

One thing I wasn't sure about how it should be modeled was the fact that `sv.always` blocks are unordered with regards to each other.  Theoretically, this means that we can merge any `sv.always` blocks with the same condition, without regard to the IR around them. I decided to rely on the type of region the `sv.always` operations were contained in. (`sv.alwaysff` ops are not `HasParent<"rtl.module">`).  For this reason, the pass skips any SSA CFG regions it finds.  I know `sv.alwaysff` blocks will always be in a Graph region, but I was thinking about how we could make this more generic with an op interface for all mergable `if`-like operations in the future.


> I would also like to generalize this into a "tidy-rtl" pass, which can do other simplifications as well

I think this is a really good idea, and what I had in mind for the future.  The original version of this code had `sv.if` statement merging as well, but it was not careful about maintaining the SSA CFG region requirements for nested `sv.if` ops (`sv.if`'s regions are SSACFG), so we decided to leave it off for another change.

One thing I was thinking is that it might make sense to implement merging as a part of a CSE pass to get better results.  In the following example, the outer if statement would be merged first, then `%b0` and `%b1` would get CSE'd, then the inner `if` could be merged (The conditional must be same SSA value for the ops to be merged):
```
if (%a) {
  %b0 = myop %c;
  if (%b0) {
  }
}
if (a) {
  %b1 = myop %c;
  if (%b1) {
  }
}
```
The builtin CSE pass has a comment about (not) CSE'ing operations with regions: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Transforms/CSE.cpp#L106.